stdlib/LinearAlgebra/src/triangular.jl

```diff
@@ -2175,6 +2339,195 @@ end
 sqrt(A::LowerTriangular) = copy(transpose(sqrt(copy(transpose(A)))))
 sqrt(A::UnitLowerTriangular) = copy(transpose(sqrt(copy(transpose(A)))))
 
+# Auxiliary functions for matrix square root
+
+# square root of upper triangular or real upper quasitriangular matrix
+function sqrt_quasitriu(A0)
+    n = checksquare(A0)
+    T = eltype(A0)
+    Tr = typeof(sqrt(real(zero(T))))
+    Tc = typeof(sqrt(complex(zero(T))))
+    if isreal(A0)
+        is_sqrt_real = true
+        if istriu(A0)
+            for i in 1:n
+                Aii = real(A0[i,i])
+                if Aii < zero(Aii)
+                    is_sqrt_real = false
+                    break
+                end
+            end
+        end
+        if is_sqrt_real
+            R = zeros(Tr, n, n)
+            A = real(A0)
+        else
+            R = zeros(Tc, n, n)
+            A = A0
+        end
+    else
+        A = A0
+        R = zeros(Tc, n, n)
+    end
+    _sqrt_quasitriu!(R, A)
+    Rc = eltype(A0) <: Real ? R : complex(R)
+    if A0 isa UpperTriangular
+        return UpperTriangular(Rc)
+    elseif A0 isa UnitUpperTriangular
+        return UnitUpperTriangular(Rc)
+    else
+        return Rc
+    end
+end
+
+function _sqrt_quasitriu!(R, A)
+    _sqrt_quasitriu_diag_block!(R, A)
+    _sqrt_quasitriu_offdiag_block!(R, A)
+    return R
+end
+
+function _sqrt_quasitriu_diag_block!(R, A)
+    n = size(R, 1)
+    ta = eltype(R) <: Complex ? complex(eltype(A)) : eltype(A)
+    i = 1
+    @inbounds while i < n
+        if iszero(A[i + 1, i])
+            R[i, i] = sqrt(ta(A[i, i]))
+            i += 1
+        else
+            # this branch is never reached when A is complex triangular
+            @views _sqrt_real_2x2!(R[i:(i + 1), i:(i + 1)], A[i:(i + 1), i:(i + 1)])
+            i += 2
+        end
+    end
+    if i == n
+        R[n, n] = sqrt(ta(A[n, n]))
+    end
+    return R
+end
+
+function _sqrt_quasitriu_offdiag_block!(R, A)
+    n = size(R, 1)
+    j = 1
+    @inbounds while j ≤ n
+        jsize_is_2 = j < n && !iszero(A[j + 1, j])
+        i = j - 1
+        while i > 0
+            isize_is_2 = i > 1 && !iszero(A[i, i - 1])
+            if isize_is_2
+                if jsize_is_2
+                    _sqrt_quasitriu_offdiag_block_2x2!(R, A, i - 1, j)
+                else
+                    _sqrt_quasitriu_offdiag_block_2x1!(R, A, i - 1, j)
+                end
+                i -= 2
+            else
+                if jsize_is_2
+                    _sqrt_quasitriu_offdiag_block_1x2!(R, A, i, j)
+                else
+                    _sqrt_quasitriu_offdiag_block_1x1!(R, A, i, j)
+                end
+                i -= 1
+            end
+        end
+        j += 2 - !jsize_is_2
+    end
+    return R
+end
+
+Base.@propagate_inbounds function _sqrt_real_2x2!(R, A)
+    a11, a21, a12, a22 = A[1, 1], A[2, 1], A[1, 2], A[2, 2]
+    θ = (a11 + a22) / 2
+    μ² = -(a11 - a22)^2 / 4 - a21 * a12
```

Comment:
The `μ` calculation seems to be vulnerable to spurious over/underflow?  Might need something similar to `hypot` here?

Similarly, it seems like one should use `hypot(θ, μ)` instead of `sqrt(θ^2 + μ²)` below?

Reply:
Addressed in #40144